lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1417 @@
+      continue;
+    if (failed(unifyAssociations(op, lhs, rhs)))
+      return failure();
+    lhs = rhs;
+  }
+  return success();
+}
+
+LogicalResult InferModuleDomains::processOp(InstanceOp op) {
+  auto module = op.getReferencedModuleNameAttr();
+  auto lookup = globals.moduleUpdateTable.find(module);
+  if (lookup != globals.moduleUpdateTable.end())
+    op = updateInstancePorts(op, lookup->second);
+  return processInstancePorts(op);
+}
+
+LogicalResult InferModuleDomains::processOp(InstanceChoiceOp op) {
+  auto module = op.getDefaultTargetAttr().getAttr();
+  auto lookup = globals.moduleUpdateTable.find(module);
+  if (lookup != globals.moduleUpdateTable.end())
+    op = updateInstancePorts(op, lookup->second);
+  return processInstancePorts(op);
+}
+
+LogicalResult InferModuleDomains::processOp(UnsafeDomainCastOp op) {
+  auto domains = op.getDomains();
+  if (domains.empty())
+    return unifyAssociations(op, op.getInput(), op.getResult());
+
+  auto input = op.getInput();
+  RowTerm *inputRow = getDomainAssociationAsRow(input);
+  SmallVector<Term *> elements(inputRow->elements);
+  for (auto domain : op.getDomains()) {
+    auto typeID = globals.circuitInfo.getDomainTypeID(domain);
+    elements[typeID] = getTermForDomain(domain);
+  }
+
+  auto *row = allocateRow(elements);
+  setDomainAssociation(op.getResult(), row);
+  return success();
+}
+
+LogicalResult InferModuleDomains::processOp(DomainDefineOp op) {
+  auto src = op.getSrc();
+  auto dst = op.getDest();
+  auto *srcTerm = getTermForDomain(src);
+  auto *dstTerm = getTermForDomain(dst);
+  if (failed(unify(dstTerm, srcTerm))) {
+    VariableIDTable idTable;
+    auto diag = op->emitOpError("failed to propagate source to destination");
+    auto &note1 = diag.attachNote();
+    note1 << "destination has underlying value: ";
+    render(note1, idTable, dstTerm);
+
+    auto &note2 = diag.attachNote(src.getLoc());
+    note2 << "source has underlying value: ";
+    render(note2, idTable, srcTerm);
+  }
+  return unify(dstTerm, srcTerm);
+}
+
+template <typename T>
+T InferModuleDomains::updateInstancePorts(T op,
+                                          const ModuleUpdateInfo &update) {
+  auto clone = op.cloneWithInsertedPortsAndReplaceUses(update.portInsertions);
+  clone.setDomainInfoAttr(update.portDomainInfo);
+  op->erase();
+  return clone;
+}
+
+template <typename T>
+LogicalResult InferModuleDomains::processInstancePorts(T op) {
+  auto circuitInfo = globals.circuitInfo;
+  auto numDomainTypes = circuitInfo.getNumDomains();
+  DenseMap<unsigned, unsigned> domainPortTypeIDTable;
+  auto domainInfo = op.getDomainInfoAttr();
+  for (size_t i = 0, e = op->getNumResults(); i < e; ++i) {
+    Value port = op.getResult(i);
+
+    LLVM_DEBUG(llvm::errs() << "handling instance port: " << port << "\n");
+
+    if (isa<DomainType>(port.getType())) {
+      auto typeID = circuitInfo.getDomainTypeID(domainInfo, i);
+      domainPortTypeIDTable[i] = typeID;
```

Comment:
Nit: Inline single-use temporaries when it doesn't affect readability:


```suggestion
      domainPortTypeIDTable[i] = circuitInfo.getDomainTypeID(domainInfo, i);
```